### PR TITLE
Import keys when same key available twice on keyserver (#1707)

### PR DESCRIPTION
**Pull request by roryflynn**
#1707 Allow for import of keys when the same key is available twice on the keyserver.

### CODE REVIEW

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java [hunk 1]
```diff
@@ -157,6 +159,50 @@ public static UncachedKeyRing decodeFromData(byte[] data)
...
+            Iterator<UncachedPublicKey> publicKeys = ring.getPublicKeys();
+
+            while (publicKeys.hasNext()) {
+                UncachedPublicKey publicKey = publicKeys.next();
+
+                boolean publicKeyMatches = (
+                        KeyFormattingUtils.convertFingerprintToHex(publicKey.getFingerprint())
+                                .equalsIgnoreCase(expectedFingerprint)
+                        || TextUtils.equals(Long.toString(publicKey.getKeyId()), expectedKeyId));
+                if (publicKeyMatches) {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java]: you can reduce nesting depth in cases like this with something like `if (!publicKeyMatches) { continue; }`, which also makes it obvious that there is no else case

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java [hunk 2]
```diff
@@ -157,6 +159,50 @@ public static UncachedKeyRing decodeFromData(byte[] data)
 
     }
 
+    public static UncachedKeyRing decodeFromData(
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java]: name could be more descriptive, i.e. `decodeFromDataFindExpected` or something like that

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java [hunk 3]
```diff
@@ -157,6 +159,50 @@ public static UncachedKeyRing decodeFromData(byte[] data)
...
+            UncachedKeyRing ring = parsed.next();
+            Iterator<UncachedPublicKey> publicKeys = ring.getPublicKeys();
+
+            while (publicKeys.hasNext()) {
+                UncachedPublicKey publicKey = publicKeys.next();
+
+                boolean publicKeyMatches = (
+                        KeyFormattingUtils.convertFingerprintToHex(publicKey.getFingerprint())
+                                .equalsIgnoreCase(expectedFingerprint)
+                        || TextUtils.equals(Long.toString(publicKey.getKeyId()), expectedKeyId));
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java]: seems better to turn expectedKeyId into a Long earlier, then compare if non-null

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/ImportOperation.java [hunk 4]
```diff
@@ -206,7 +206,7 @@ private ImportKeyResult serialKeyRingImport(Iterator<ParcelableKeyRing> entries,
                                 log.add(LogType.MSG_IMPORT_FETCH_KEYSERVER, 2, entry.mKeyIdHex);
                                 data = keyServer.get(entry.mKeyIdHex).getBytes();
                             }
-                            key = UncachedKeyRing.decodeFromData(data);
+                            key = UncachedKeyRing.decodeFromData(data, entry.mExpectedFingerprint, entry.mKeyIdHex);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/ImportOperation.java]: what about the other calls to decodeFromData?

**Comment by roryflynn** [comment]
@Valodim should be ready for another review!

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java [hunk 5]
```diff
@@ -163,6 +163,54 @@ public static UncachedKeyRing decodeFromData(byte[] data)
...
+                UncachedPublicKey publicKey = publicKeys.next();
+
+                boolean fingerprintMatches = KeyFormattingUtils.convertFingerprintToHex(publicKey.getFingerprint())
+                        .equalsIgnoreCase(expectedFingerprint);
+                boolean keyIdMatches = (expectedKeyIdLong != null
+                        && expectedKeyIdLong.longValue() == publicKey.getKeyId());
+
+                boolean publicKeyMatches = ( fingerprintMatches || keyIdMatches );
+
+                if ( ! publicKeyMatches) { continue; }
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java]: code style

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/util/KeyFormattingUtils.java [hunk 6]
```diff
@@ -271,6 +271,25 @@ private static String convertKeyIdToHex32bit(long keyId) {
         return hexString;
     }
 
+    /**
+     * Convert key id from hex string to long
+     * @param keyIdHex - 64 bit hex string representation of keyId, including "0x" prefix
+     * @return keyId as long
+     */
+    public static Long convertKeyIdHexToLong(String keyIdHex) {
+        if (keyIdHex == null) { return null;}
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/util/KeyFormattingUtils.java]: code style

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/util/KeyFormattingUtils.java [hunk 7]
```diff
@@ -271,6 +271,25 @@ private static String convertKeyIdToHex32bit(long keyId) {
...
+
+        //we can't convert directly from hex
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/util/KeyFormattingUtils.java]: why not?

Comment (roryflynn) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/util/KeyFormattingUtils.java]: It appears you can't parse a long from a two's complement encoded hex string using Long.parseLong or Long.decode http://stackoverflow.com/questions/1410168/how-to-parse-negative-long-in-hex-in-java

I can make the comment a bit more verbose if you want?

Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/util/KeyFormattingUtils.java]: sounds fine. including the link might also be a good idea

**Comment by roryflynn** [comment]
@Valodim Code style issues + comments should be fixed now

**Comment by Valodim** [comment]
Would you add a positive and a negative unit test for the behavior you introduced? That would be awesome :+1:

**Comment by dschuermann** [comment]
@roryflynn Still working on this PR?

**Comment by roryflynn** [comment]
@dschuermann Apologies, not at the minute due to exams. Others are welcome to finish it off by creating the test cases though. If not, I'll try to finish it in about two weeks time.

**Comment by dschuermann** [comment]
@roryflynn Take your time with the exams. We appreciate if you get back to us afterwards :)

**Comment by Valodim** [comment]
closing as stale

**Pull request closed without merging**